File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.cpp
class Solution {
public:
    int minSetSize(vector<int>& arr) {
        int n = arr.size();
        
        unordered_map<int,int> mapka;
        for(auto& val : arr){
            mapka[val]++;
        }
        
        vector<int> counting(n + 1);
        for(auto&[key , val] : mapka){
            counting[val]++;
        }
        
        int ans = 0 , removed = 0 , half = n/2 , freq = n;
        while(removed < half){
            while(counting[freq] == 0)
                freq--;
            
            ans++;
            removed += freq;
            counting[freq]--;
        }
        
        return ans;
    }
};

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution {
    public int minSetSize(int[] arr) {
       List<Integer> list = Arrays.stream(arr).boxed().collect(Collectors.toList());
        HashMap<Integer, Integer> freqMap = new HashMap<>();
        for (Integer integer : list) {
            freqMap.put(integer, freqMap.getOrDefault(integer, 0) + 1);
        }

        list = list.stream().sorted(Comparator.comparing(freqMap::get).reversed()).collect(Collectors.toList());

        int removed = 0 , half= list.size() / 2;
        int ans = 0;
        while(removed < half){
            ans++;
            removed += freqMap.get(list.get(removed));
        }

        return ans;
    }
}